proposals/csharp-11.0/low-level-struct-improvements.md

```diff
@@ -869,6 +869,194 @@ ref struct S<T>
+    - `ref<$a> (T<$b>)expr` the value lifetime is `$b` for `T<...>` and the ref lifetime is `$a`.
+
+Next let's define the rules that allow us to map C# syntax to the underlying model.
+
+The lifetime parameter `$this` on type definitions is _not_ predefined but it does have a few rules associated with it when it is defined:
+- It must be the first lifetime parameter.
+- It must be covariant: `out $this`. 
+- The lifetime parameters of all non-ref fields, and the ref lifetime of ref fields, must be `$this`
+
+For brevity sake a type which has no explicit lifetime parameters treated as if there is `out $this` defined and applied to all fields of the type. A type with a `ref` field must define explicit lifetime parameters.
+
+These rules exists to support our existing invariant that `T` can be assigned to `scoped T` for all types. That maps down to `T<$a, ...>` being assignable to `T<$local, ...>` for all lifetimes known to be a subtype of `$local`. Further this supports other items like being able to assign `Span<T>` from the heap to those on the stack. This does exclude types where fields have differing lifetimes for non-ref values but that is the reality of C# today. Changing that would require a significant change of C# rules that would need to be mapped out. 
+
+The type of `this` for a type `S<out $this, ...>` inside an instance method is implicitly defined as the following:
+- For normal instance method: `ref<$local> S<$ro, ...>`
+- For instance method annotated with `[UnscopedRef]`: `ref<$ro> S<$ro, ...>`
+The lack of an explicit `this` parameter forces the implicit rules here. For complex samples and discussions likely better to use an explicit parameter.
+
+The C# method syntax maps to the model in the following ways: 
+
+- `ref` parameters have a ref lifetime of `$ro`
+- parameters of type `ref struct` have a this lifetime of `$cm`
+- ref returns have a ref lifetime of `$ro`
+- returns of type `ref struct` have a value lifetime of `$ro`
+- `scoped` on a parameter or `ref` changes the lifetime to be `$local`
+
+Given that let's explore a simple example that demonstrates the model here: 
+
+```csharp
+ref int M1(ref int i) => ...
+
+// Maps to the following. 
+
+ref<$ro> int Identity<$ro>(ref<$ro> int i)
+{
+    // okay: has ref lifteime $ro which is equal to $ro
+    return ref i;
+
+    // okay: has ref lifetime $heap which is >= $ro
+    int[] array = new int[42];
+    return ref array[0];
+
+    // error: has ref lifetime $local which has no relationship to $a hence 
+    // it's illegal
+    int local = 42;
+    return ref local;
+}
+```
+
+Now let's explore the same example using a `ref struct`: 
+
+```csharp
+ref struct S
+{
+    ref int Field;
+
+    S(ref int f)
+    {
+        Field = ref f;
+    }
+}
+
+S M2(ref int i, S span1, scoped S span2) => ...
+
+// Maps to 
+
+ref struct S<out $this>
+{
+    // Implicitly 
+    ref<$this> int Field;
+
+    S<$ro> new<$ro>(ref<$ro> int f)
+    {
+        Field = ref f;
+    }
+}
```

Comment:
Are you sure this is what you meant? I would have expected

```suggestion
ref struct S<out $this, out $ro>
{
    // Implicitly 
    ref<$ro> int Field;

    S<$this, $ro> new<$this, $ro>(ref<$ro> int f)
    {
        Field = ref f;
    }
}
```

Basically, `$this` would be the lifetime of the fields themselves, when a ref is taken to them, but the lifetime of a ref field could be anything, as long as it's at least as long as `$this`.

Reply:
No, it's meant as it's written. One idea to keep in mind is that types only have one level of _interesting_ lifetime today. Specifically that is the lifetime of the value. Yes we have ref fields but the lifetime of the _value_ of the `ref` field is always `$heap`. Having the lifetime of a field, be it a normal or ref field, be different than the containing type isn't possible today. So anytime you find yourself trying to describe an existing type with two lifetimes you're likely going off the wrong path here. 

I agree that in the future it may be valuable to let more flexibility like this into the model but there is no syntax to support such a construct today.

Reply:
Not sure I follow.

If I have the following method:

```C#
ref struct S
{
    int field;
    ref int refField;

    void M()
    {
        ref int x = ref field;
        ref int y = ref refField;
    }
}
```

What are the lifetimes of `x` and `y`?

Reply:
Rewriting `M` as a static method to facility an explicit `this` parameter that makes it easier to discuss. 

```csharp
ref struct S<out $this>
{
    int field;
    ref<$this> int refField;

    static M<$this>(ref<$local> S<$ro> @this)
    {
        ref<$local> int x = ref field;
        ref<$ro> int y = ref refField;
    }
}
```

These follow from the rules for `ref` expressions earlier in the docs

> - The `ref` of a variable has a lifetime of the 
>    - For a ref local, parameter, field or return of type `ref<$a> T` the lifetime is `$a`
>    - `$heap` for all reference types and fields of reference types
>    - `$local` for everything else

Reply:
Got it, so you're really kind of smuggling what I would call the `$this` lifetime along with the parameter itself, as part of the parent `ref` -- that makes sense!

In that case, my only piece of feedback is that I was a bit confused by the name "$this", since I thought of that as the lifetime of the struct itself, but it's really the lifetime of the ref fields stored inside the struct.

Reply:
It's also the lifetime of normal `ref struct` fields as well. For example you'd really have the following:

```csharp
ref struct S<out $this> {
  Span<$this, int> Span;
  ref<$this> int refField;
}
```